oprf/oprf.go

```diff
@@ -0,0 +1,277 @@
+//
+// Package oprf provides an implementation of Oblivious Pseudorandom Functions
+// (OPRFs), as defined on draft-irtf-cfrg-voprf.
+// It implements:
+// For a Client:
+//   - Blind
+//   - Unblind
+//   - Finalize
+//
+// For a Server:
+//   - Setup
+//   - Evaluate
+//   - VerifyFinalize
+// References
+//  - OPRF draft: https://datatracker.ietf.org/doc/draft-irtf-cfrg-voprf/
+package oprf
+
+import (
+	"crypto/subtle"
+	"errors"
+
+	"github.com/cloudflare/circl/oprf/group"
+)
+
+// SuiteID is a type that represents the ID of a Suite.
+type SuiteID uint16
+
+const (
+	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group.
+	OPRFP256 SuiteID = 0x0003
+	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group.
+	OPRFP384 SuiteID = 0x0004
+	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group.
+	OPRFP521 SuiteID = 0x0005
+)
+
+var (
+	// OPRFMode is the context string to define a OPRF.
+	OPRFMode byte = 0x00
+)
+
+var (
+	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	pubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+}
+
+// Server is a representation of a Server during protocol execution.
+type Server struct {
+	suite *group.Ciphersuite
+	ctx   []byte
+	Kp    *KeyPair
+}
+
+func generateCtx(id SuiteID) []byte {
```

Comment:
nit: can we spell out `Context` rather than abbreviate to `Ctx`?

Reply:
Sure.